osu.Game/Screens/Spectate/SpectatorScreen.cs

```diff
@@ -0,0 +1,237 @@
+                    if (beatmapSet.Beatmaps.Any(b => b.OnlineBeatmapID == state.BeatmapID))
+                        updateGameplayState(userId);
+                }
+            }
+        }
+
+        private void userBeganPlaying(int userId, SpectatorState state)
+        {
+            if (state.RulesetID == null || state.BeatmapID == null)
+                return;
+
+            lock (stateLock)
+            {
+                if (!userMap.ContainsKey(userId))
+                    return;
+
+                spectatorStates[userId] = state;
+                OnUserStateChanged(userId, state);
+
+                updateGameplayState(userId);
+            }
+        }
+
+        private void updateGameplayState(int userId)
+        {
+            lock (stateLock)
+            {
+                Debug.Assert(userMap.ContainsKey(userId));
+
+                var spectatorState = spectatorStates[userId];
+                var user = userMap[userId];
+
+                var resolvedRuleset = rulesets.AvailableRulesets.FirstOrDefault(r => r.ID == spectatorState.RulesetID)?.CreateInstance();
+                if (resolvedRuleset == null)
+                    return;
+
+                var resolvedBeatmap = beatmaps.QueryBeatmap(b => b.OnlineBeatmapID == spectatorState.BeatmapID);
+                if (resolvedBeatmap == null)
+                    return;
+
+                var score = new Score
+                {
+                    ScoreInfo = new ScoreInfo
+                    {
+                        Beatmap = resolvedBeatmap,
+                        User = user,
+                        Mods = spectatorState.Mods.Select(m => m.ToMod(resolvedRuleset)).ToArray(),
+                        Ruleset = resolvedRuleset.RulesetInfo,
+                    },
+                    Replay = new Replay { HasReceivedAllFrames = false },
+                };
+
+                var gameplayState = new GameplayState(score, resolvedRuleset, beatmaps.GetWorkingBeatmap(resolvedBeatmap));
+
+                gameplayStates[userId] = gameplayState;
+                StartGameplay(userId, gameplayState);
+            }
+        }
+
+        private void userSentFrames(int userId, FrameDataBundle bundle)
+        {
+            lock (stateLock)
+            {
+                if (!userMap.ContainsKey(userId))
+                    return;
+
+                if (!gameplayStates.TryGetValue(userId, out var gameplayState))
+                    return;
+
+                // The ruleset instance should be guaranteed to be in sync with the score via ScoreLock.
+                Debug.Assert(gameplayState.Ruleset != null && gameplayState.Ruleset.RulesetInfo.Equals(gameplayState.Score.ScoreInfo.Ruleset));
+
+                foreach (var frame in bundle.Frames)
+                {
+                    IConvertibleReplayFrame convertibleFrame = gameplayState.Ruleset.CreateConvertibleReplayFrame();
+                    convertibleFrame.FromLegacy(frame, gameplayState.Beatmap.Beatmap);
```

Comment:
While refactoring this, I also noticed that this line is completely wrong. The logic hasn't changed but the method expects the playable beatmap at this point (see: `ManiaReplayFrame`).

It's just good to know at this point - I'm unsure exactly how to handle this yet, but it'll be fixed in some future PR.